src/psyclone/parse/file_info.py

```diff
@@ -77,27 +141,315 @@ def filename(self):
-        :rtype: str
 
         '''
-        if self._source_code is None:
+        if self._source_code:
+            return self._source_code
+
+        if verbose:
+            # TODO #11: Use logging for this
+            print(
+                f"- Source file '{self._filename}': "
+                f"Loading source code"
+            )
+
+        try:
             # Specifying errors='ignore' simply skips any characters that
             # result in decoding errors. (Comments in a code may contain all
             # sorts of weird things.)
-            with open(self._filename, "r", encoding='utf-8',
-                      errors='ignore') as file_in:
+            with open(
+                self._filename, "r", encoding="utf-8", errors="ignore"
+            ) as file_in:
                 self._source_code = file_in.read()
+        except FileNotFoundError as err:
+            raise FileNotFoundError(
+                f"No such file or directory '{self._filename}'."
+            ) from err
+
+        # Compute hash sum which will be used to check cache of fparser tree
+        self._source_code_hash_sum = hashlib.md5(
+            self._source_code.encode()
+        ).hexdigest()
 
         return self._source_code
+
+    def _cache_load(
+        self,
+        verbose: bool = False,
+    ) -> _CacheFileInfo:
+        """Load fparser parse tree from the cache file if possible.
+        This also checks for matching checksums after loading the data
+        from the cache.
+        The checksum is based solely on a hashsum of the source code itself,
+        see code below.
+
+        :param verbose: Produce some verbose output
+        """
+
+        if not self._use_caching:
+            return
+
+        # Load the source code in case it's not yet loaded.
+        # This also fills in the hash sum
+        self.get_source_code()
+
+        # Check whether cache was already loaded
+        if self._cache_data_load is not None:
+            return self._cache_data_load
+
+        # Load cache file.
+        # Warning: There could be race conditions, e.g., in parallel builds.
+        # In the worst case some content is read which is incomplete or
+        # basically garbage. This will lead either to an Exception from the
+        # unpickling or a non-matching checksum which is both caught below.
+        try:
+            filehandler = open(self._filepath_cache, "rb")
+        except FileNotFoundError:
+            if verbose:
+                # TODO #11: Use logging for this
+                print(f"  - No cache file '{self._filepath_cache}' found")
+            return None
+
+        # Unpack cache file
+        try:
+            cache: _CacheFileInfo = pickle.load(filehandler)
+        except Exception as ex:
+            print(f"  - Error while reading cache file - ignoring: {str(ex)}")
+            return None
+
+        # Verify checksums
+        if cache._source_code_hash_sum != self._source_code_hash_sum:
```

Comment:
@arporter It turned out that already no `self._cache_data_load` is generated at all if the checksums aren't matching.